tests/extmod/deflate_compress_memory_error.py

```diff
@@ -0,0 +1,39 @@
+# Test deflate.DeflateIO compression, with out-of-memory errors.
+
+try:
+    # Check if deflate is available.
+    import deflate
+    import io
+except ImportError:
+    print("SKIP")
+    raise SystemExit
+
+# Check if compression is enabled.
+if not hasattr(deflate.DeflateIO, "write"):
+    print("SKIP")
+    raise SystemExit
+
+# Create a compressor object.
+b = io.BytesIO()
+g = deflate.DeflateIO(b, deflate.RAW, 15)
+
+# Then, use up most of the heap.
+l = []
+while True:
+    try:
+        l.append(bytearray(1000))
+    except:
+        break
+l.pop()
+
+# Try to compress.  This will try to allocate a large window and fail.
```

Comment:
Could you do this test by locking the heap before calling `g.write()`, instead? (My quick read of moddeflate doesn't show up any temporarily allocation that would fail first, but maybe one exists?)

Reply:
That's a nice idea, but it won't work: the existing code (on master) that has a crash does a small alloc first of `mp_obj_deflateio_write_t` and then the large alloc of the window.  You need the first to pass -- so the code sets `self->write` -- and then the second to fail.  In that case the object is in a bad state, with `self->write` non-NULL yet the window unallocated.

Reply:
I see, I was only thinking about the test in this PR branch where it does the large allocation first so this would fail now. I agree it's not strictly testing the same thing because on master it won't trigger the bug, though.

Reply:
It's a tricky one.

I guess the best way to look at is this test prevents a regression.  Eg if someone comes along and reorders the logic here to allocate the window later on in the function, that test will fail.